### ISSUE TITLE
 ERROR: Failed to get certificate (website certificate is OK)

### ISSUE THREAD

**Issue by M4rt1n12** [selivan/https-ssl-cert-check-zabbix#44]
### Environment

`openssl version` says:

```
OpenSSL 1.1.1f  31 Mar 2020
```

Operating System: Ubuntu Server 20.04

### Expected Behavior

Remaining Lifetime of certificate

### Current Behavior

-65535
ERROR: Failed to get certificate

### Diagnostics

Test results from https://www.whynopadlock.com

```
Tested URL
XXX
Test completed
Tue, Mar 7, 2023 11:57 AM Eastern Time (GMT -5)
Results URL

SSL Connection - Pass
SSL Certificate Info

Certificate Issuer
    Let's Encrypt
Certificate Type
    R3
Issued On
    2023-03-05 

Force HTTPS
Your webserver is forcing the use of SSL.
Valid Certificate
Your SSL Certificate is installed correctly.
Domain Matching
Your SSL certificate matches your domain name!
Protected Domains:

XXX (10 domains)

Signature
Your SSL certificate is using a sha256WithRSAEncryption signature!
Expiration Date
Your SSL certificate is current. Your SSL certificate expires in 87 days. (2023-06-03)
Mixed Content - Pass
You have no mixed content.

```

**Comment by selivan** [comment]
Hey. Sorry for the delayed answer, I missed the issue.

Most likely server where you are running the script fails to connect to tested host and port, while server running whynopadlock.com connects successfully.

Could you please post output of 

```
echo | openssl s_client -servername <tls domain name> -connect <host>:<port>
```

If that command succeeds, but the script fails, then it is an error in the script.

**Comment by M4rt1n12** [comment]
The output is valid, so the script is wrong. I simply generated a new one with chatgpt:

`#!/bin/bash

# Get the website from the command line argument
WEBSITE=$1
PORT=443

# Get the expiration date of the SSL certificate
expiration_date=$(echo | openssl s_client -servername $WEBSITE -connect $WEBSITE:$PORT 2>/dev/null | openssl x509 -noout -dates | grep notAfter | awk -F "=" '{print $2}')

# Convert the expiration date to epoch timestamp
expiration_timestamp=$(date -d "$expiration_date" +%s)

# Get the current timestamp
current_timestamp=$(date +%s)

# Calculate the remaining days until expiration
remaining_days=$((($expiration_timestamp - $current_timestamp) / (60 * 60 * 24)))

# Output the remaining days
echo $remaining_days
`

**Comment by M4rt1n12** [comment]
The Output is OK. But I think it's an error in the deployment of the certificate. I used it for a filemaker webserver and after generating it, I manually imported it to the filemaker webserver. Now it there's a different expiration date in the browser and in the filemaker admin console. I guess there is an error in the deployment.

**Comment by selivan** [comment]
If `s_client` output is OK, but the script gives `ERROR: Failed to get certificate`, then it is a bug in a script. So I have to see the output to debug it. Domain name and other sensitive information redacted, of course.

**Comment by Amblikmees** [comment]
I used basically the same command what can be found in script
```
if ! output=$( echo \
| timeout "$check_timeout" openssl s_client $starttls $starttls_proto -servername "$domain" -verify_hostname "$domain" -connect "$host":"$port" $tls_version $s_client_options 2>/dev/null )
```
like this:
`$ echo | timeout 5 openssl s_client -connect demo.test.mydomain.com:443 2>/dev/null`
And the output was following:
```
CONNECTED(00000003)
---
Certificate chain
 0 s:/C=xx/O=xx/OU=xx
   i:/DC=x/DC=xx/CN=xx
 1 s:/DC=x/DC=xx/CN=xx
   i:/DC=x/DC=xx/CN=xx
---
Server certificate
-----BEGIN CERTIFICATE-----
....
-----END CERTIFICATE-----
subject=/C=xx/O=xx/OU=xx
issuer=/DC=xx/DC=xx/CN=xx
---
Acceptable client certificate CA names
..
Client Certificate Types: RSA sign, DSA sign, ECDSA sign
Requested Signature Algorithms: RSA+SHA512:DSA+SHA512:ECDSA+SHA512:RSA+SHA384:DSA+SHA384:ECDSA+SHA384:RSA+SHA256:DSA+SHA256:ECDSA+SHA256:RSA+SHA224:DSA+SHA224:ECDSA+SHA224:RSA+SHA1:DSA+SHA1:ECDSA+SHA1
Shared Requested Signature Algorithms: RSA+SHA512:DSA+SHA512:ECDSA+SHA512:RSA+SHA384:DSA+SHA384:ECDSA+SHA384:RSA+SHA256:DSA+SHA256:ECDSA+SHA256:RSA+SHA224:DSA+SHA224:ECDSA+SHA224:RSA+SHA1:DSA+SHA1:ECDSA+SHA1
Peer signing digest: SHA512
Server Temp Key: ECDH, P-256, 256 bits
---
SSL handshake has read 4989 bytes and written 138 bytes
---
New, TLSv1/SSLv3, Cipher is ECDHE-RSA-AES256-GCM-SHA384
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-RSA-AES256-GCM-SHA384
    Session-ID:
    Session-ID-ctx:
    Master-Key: xxx
    Key-Arg   : None
    Krb5 Principal: None
    PSK identity: None
    PSK identity hint: None
    Start Time: 1685621603
    Timeout   : 300 (sec)
    Verify return code: 10 (certificate has expired)
---
```
And i was able to get expiration date/number by doing manually get_expire_days() function part. Somehow the IF statement think its False.

**Comment by selivan** [comment]
@Amblikmees seems that `s_client` returns non-zero code. Could you show what `openssl version` says on that system?

Also, I tried to make a work around here: https://github.com/selivan/https-ssl-cert-check-zabbix/blob/issue-44/ssl_cert_check.sh

Could you verify if it works?

**Comment by Amblikmees** [comment]
Openssl version: OpenSSL 1.0.2k-fips
And the work around worked successfully!

**Comment by selivan** [comment]
@Amblikmees fix merged to master branch: f663584
